src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h

```diff
@@ -1186,6 +1186,42 @@ namespace UiaOperationAbstraction
             return *std::get<LocalType>(m_member);
         }
 
+        UiaBool operator==(const UiaArray& rhs) const
+        {
+            if (ShouldUseRemoteApi())
+            {
+                auto mutableThis = *this;
+                mutableThis.ToRemote();
+
+                auto mutableRhs = rhs;
+                mutableRhs.ToRemote();
+                return std::get<RemoteType>(mutableThis.m_member).IsEqual(std::get<RemoteType>(mutableRhs.m_member));
+            }
+
+            auto localVector = std::get<LocalType>(m_member);
+            auto localRhsVector = std::get<LocalType>(rhs.m_member);
+
+            if (localVector->size() != localRhsVector->size())
+            {
+                return false;
+            }
+
+            for (size_t i = 0; i < localVector->size(); ++i)
+            {
+                if (localVector->at(i) != localRhsVector->at(i))

```

Comment:
Does `operator!=` consistently do the right thing here for types that we can have in this `vector`? For instance, what is the type here if we have `UiaArray<UiaElement>` and what does `operator!=` do?

Reply:
Thanks for catching this, this won't do the right comparison for types such UiaElements or UiaTextRange.
I changed this to only compare with supported types, same as the our internal implementation, when it is not one those supported types, it throws exception. I thought about use static assert to catch it at compile time, but I figure maybe it is better to make it consistent with our remote operation implementation which throws exception(E_UNEXPECTED) at runtime. Let me know what is your take on this or better suggestions here.

Reply:
If we can detect an invalid template instantiation at compile time, we should do so. If we already know for sure that a particular usage will lead to a runtime error, we should prevent folks from writing what would clearly be a bug. The idea behind the abstraction wrapper is exactly to provide a more convenient API for users that are able to use C++ (and want to use the abstraction) and solve some of the issues that cannot be solved in the platform (or that would be too constricting or hinder future development if put in the platform API).

(As a related example, while in the underlying Remote Ops VM arrays can contain any operand type, the abstraction makes it more strongly typed for convenience and ensures that we're only putting elements of a particular type -- as long as we're staying within the bounds of the abstraction API. This situation is similar, IMV.)

Reply:
Thanks for the explanation. Changes it to use static assert now.